src/catalog/src/durable/persist.rs

```diff
@@ -1023,42 +1214,62 @@ impl UnopenedPersistCatalogState {
         bootstrap_args: &BootstrapArgs,
     ) -> Result<Box<dyn DurableCatalogState>, CatalogError> {
         self.mode = mode;
+
+        // Validate the current deploy generation.
+        match (&self.mode, &self.fenceable_token) {
+            (_, FenceableToken::Unfenced { .. } | FenceableToken::Fenced { .. }) => {
+                return Err(DurableCatalogError::Internal(
+                    "catalog should not have fenced before opening".to_string(),
+                )
+                .into());
+            }
+            (
+                Mode::Writable | Mode::Savepoint,
+                FenceableToken::Initializing {
+                    current_deploy_generation: None,
+                    ..
+                },
+            ) => {
+                return Err(DurableCatalogError::Internal(format!(
+                    "cannot open in mode '{:?}' without a deploy generation",
+                    self.mode,
+                ))
+                .into());
+            }
+            _ => {}
+        }
+
         let read_only = matches!(self.mode, Mode::Readonly);
 
         // Fence out previous catalogs.
-        self.sync_to_current_upper().await?;
-        let prev_epoch = self.epoch.validate()?;
-        let mut fence_updates = Vec::with_capacity(2);
-        if let Some(prev_epoch) = prev_epoch {
-            fence_updates.push((StateUpdateKind::Epoch(prev_epoch), -1));
-        }
-        let mut current_epoch = prev_epoch.unwrap_or(MIN_EPOCH).get();
-        // Only writable catalogs attempt to increment the epoch.
-        if matches!(self.mode, Mode::Writable) {
-            current_epoch = current_epoch + 1;
-        }
-        let current_epoch = Epoch::new(current_epoch).expect("known to be non-zero");
-        fence_updates.push((StateUpdateKind::Epoch(current_epoch), 1));
-        let current_epoch = FenceableToken::new(Some(current_epoch), FenceError::epoch);
-        debug!(
-            ?self.upper,
-            ?prev_epoch,
-            ?current_epoch,
-            "fencing previous catalogs"
-        );
-        self.epoch = current_epoch;
-        if matches!(self.mode, Mode::Writable) {
-            loop {
+        loop {
```

Comment:
Is it safe to move this loop higher because the call to `generate_unfenced_token(...)` doesn't update any internal state?

Reply:
Yeah exactly